src/kubernetes_cluster/proof/kubernetes_api_liveness.rs

```diff
@@ -223,6 +246,182 @@ pub open spec fn no_req_before_rest_id_is_in_flight(rest_id: RestId) -> StatePre
     }
 }
 
+/// Newly created message means it doesn't exist in in_flight in last state but exists in the current state.
+pub open spec fn every_new_in_flight_req_msg_satisfies(requirements: FnSpec(Message) -> bool) -> ActionPred<Self> {
+    |s: Self, s_prime: Self| {
+        forall |msg: Message|
+            !s.message_in_flight(msg)
+            && #[trigger] s_prime.message_in_flight(msg)
+            && msg.dst.is_KubernetesAPI()
+            && msg.content.is_APIRequest()
+            ==> requirements(msg)
+    }
+}
+
+pub open spec fn every_in_flight_req_msg_satisfies(requirements: FnSpec(Message) -> bool) -> StatePred<Self> {
+    |s: Self| {
+        forall |msg: Message|
+            #[trigger] s.message_in_flight(msg)
+            && msg.dst.is_KubernetesAPI()
+            && msg.content.is_APIRequest()
+            ==> requirements(msg)
+    }
+}
+
+/// This lemma shows that if spec ensures every newly created Kubernetes api request message satisfies some requirements,
+/// the system will eventually reaches a state where all Kubernetes api request messages satisfy those requirements.
+///
+/// To require "every newly create Kubernetes api request message satisfies some requirements", we use a FnSpec (i.e., a closure)
+/// as parameter which can be defined by callers and require spec |= [](every_new_in_flight_req_msg_satisfies(requirements)).
+/// Note that in the precondition, we don't actually require always for this predicate, because the spec need to be stable and the always
+/// one can be derived. This also alleviate the preliminary work for the callers to use this lemma.
+///
+/// The last parameter must be equivalent to every_in_flight_req_msg_satisfies(requirements)
+pub proof fn lemma_true_leads_to_always_every_in_flight_req_msg_satisfies(
+    spec: TempPred<Self>, requirements: FnSpec(Message) -> bool, msg_state_pred: StatePred<Self>
+)
+    requires
+        valid(stable(spec)),
+        spec.entails(lift_action(Self::every_new_in_flight_req_msg_satisfies(requirements))),
```

Comment:
I am trying to understand this precondition: what you later need is actually `spec.entails(always(lift_action(Self::every_new_in_flight_req_msg_satisfies(requirements))))` but `valid(stable(spec))` can easily give you that so you don't require the "always" here, is that right?

Reply:
Yes. We can also remove `valid(stable(spec))` and add `[]` to the second. If you want to use `strengthen_next_n`, the alternative is better.

Reply:
I also like this lemma a lot. The only counter-intuitive part is actually that this particular precondition does not have "always". I think it would be a lot easier to make sense of it if we just require the always here (and then we no longer need to require stable spec I guess?)

> This also alleviate the preliminary work for the callers to use this lemma.

I understand this but it seems to me proving the always is very simple (just one lemma call), especially if we later use the invariant_action lemma I was proposing.

Reply:
So I wrote this lemma before you proposed everything.... You have to combine all the always...

Reply:
Yes, understand. I have some race conditions here. When I am writing the comment I don't always see your newest comments. If you plan to do the refactoring mentioned above pls go ahead, otherwise let's put it in the management file as you suggested above.

Reply:
I made some changes (you can take a look). But I don't write the `invariant_action` because it's very obvious.